### PR TITLE
Redux toolkit immer typescript

### PR DESCRIPTION
**Pull request by pmyip3**

### CODE REVIEW

File: src/component/TodoItem.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+interface ParamTypes {
+    todoId: string
+}
+
+
+function TodoItem(){
+    const { todoId } = useParams<ParamTypes>()
+    console.log(todoId)
+    const[redirect, setRedirect] = useState(false)  
+    const todoItemList = useSelector((state: RootState) => state)
```
Comment (k3rw1n-weng) [inline src/component/TodoItem.tsx:17]: useSelector<TodoItem>((state: RootState) => state[todoId]);

Comment (k4w21) [inline src/component/TodoItem.tsx:17]: const todoItem = useSelector((state: RootState) =>  state[todoId])

File: src/component/TodoItem.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+
+function TodoItem(){
+    const { todoId } = useParams<ParamTypes>()
+    console.log(todoId)
+    const[redirect, setRedirect] = useState(false)  
+    const todoItemList = useSelector((state: RootState) => state)    
+    const itemListObject: ITodoItemObject[] = Object.values(todoItemList)
+
+    const TodoItem = itemListObject.filter(item => item.id === todoId)
```
Comment (k4w21) [inline src/component/TodoItem.tsx:20]: case

File: src/interface/models.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export interface ITodoItem {
+    id: string
+    description: string
+    category: string
+    content: string
+}
+
+export interface ITodoItemObject {
+    id: string
+    todoItem: ITodoItem
```
Comment (k4w21) [inline src/interface/models.tsx:10]: Over wrapping

File: src/component/InputForm.tsx [hunk 4]
```diff
@@ -37,7 +41,12 @@ function InputForm(props){
...
                     type="button" 
                     onClick={()=>{
-                        props.addTodo({description,category, content})
+                        const id = uuidv1();
+                        console.log(id, description,category, content)
+                        const todoItem = {
+                            id, description, category, content
+                        }
+                        dispatch(addTodo({id, todoItem}))
                         setDescription("")
```
Comment (k4w21) [inline src/component/InputForm.tsx:50]: multiple refresh

File: src/component/TodoItem.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    console.log(TodoItem)
+
+    function renderRedirect(){
+        if (redirect){
+            return <Redirect to='/todo'/>
```
Comment (k4w21) [inline src/component/TodoItem.tsx:26]: useHistory

File: src/redux/index.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+export const deleteAllTodos = createAction<ITodoItemObject[]>('delete_all_todo')
+
+const initialTodoList = {}
+
+const reducer = createReducer(initialTodoList, {
+    [addTodo.type]: produce((draft, action) => {
+        draft[action.payload.id] = action.payload
+    }),
+    [deleteTodo.type]: produce((draft, action) => {delete draft[action.payload]}),
+    [deleteAllTodos.type]: produce((draft, action) => {action.payload.map((todo: { id: string }) => delete draft[todo.id])}) //don't know how to empty an object
```
Comment (k4w21) [inline src/redux/index.tsx:16]: Object.keys(draft).forEach(key => { delete draft[key] })

File: src/redux/index.tsx [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import { ITodoItem, ITodoItemObject} from '../interface/models'
+
+export const addTodo = createAction<ITodoItemObject>('add_todo')
+export const deleteTodo = createAction<string>('delete_todo')
+export const deleteAllTodos = createAction<ITodoItemObject[]>('delete_all_todo')
+
+const initialTodoList = {}
+
+const reducer = createReducer(initialTodoList, {
+    [addTodo.type]: produce((draft, action) => {
```
Comment (k4w21) [inline src/redux/index.tsx:12]: produce is implied when using createReducer

File: src/App.tsx [hunk 8]
```diff
@@ -11,7 +11,7 @@ function App() {
       <NavBar />
       <Switch>
         <Route exact path="/todo"> <TodoPage /></Route>
-        <Route exact path="/todo/:todoIndex"> <TodoItem /></Route>
+        <Route exact path="/todo/:todoId"> <TodoItem /></Route>
         <Route exact path="/about"> <AboutPage /></Route>
```
Comment (k4w21) [inline src/App.tsx:15]: Adding redirect for non-match path

File: src/component/InputForm.tsx [hunk 9]
```diff
@@ -1,9 +1,13 @@
 import React, {useState} from 'react'
+import { v1 as uuidv1 } from 'uuid'
+import { useDispatch } from 'react-redux'
+import { addTodo } from '../redux'
 
-function InputForm(props){ 
-    const [description, setDescription] = useState("")
-    const [category, setCategory] = useState("")
-    const [content, setContent] = useState("")
+function InputForm(){
```
Comment (k4w21) [inline src/component/InputForm.tsx:6]: Start using arrow function

File: src/component/InputForm.tsx [hunk 10]
```diff
@@ -37,7 +41,12 @@ function InputForm(props){
                 <button 
                     type="button" 
                     onClick={()=>{
-                        props.addTodo({description,category, content})
+                        const id = uuidv1();
+                        console.log(id, description,category, content)
+                        const todoItem = {
```
Comment (k4w21) [inline src/component/InputForm.tsx:46]: no need over wrapping

File: src/App.tsx [hunk 11]
```diff
@@ -1,18 +1,19 @@
...
+const App = () => {
   return (
     <React.Fragment> 
       <NavBar />
       <Switch>
         <Route exact path="/todo"> <TodoPage /></Route>
-        <Route exact path="/todo/:todoId"> <TodoItem /></Route>
+        <Route exact path="/todo/:todoId"> <TodoItemDetail /></Route>
         <Route exact path="/about"> <AboutPage /></Route>
+        <Route render={() => <Redirect to="/" />} />
```
Comment (hunteva) [inline src/App.tsx:16]: <Redirect to="/todo" />

File: src/component/InputForm.tsx [hunk 12]
```diff
@@ -15,15 +13,21 @@ function InputForm(){
...
+                        }}
                     />
                 </h2>
                 <h2>
                     Category: 
                     <select 
-                        value={category}
-                        onChange={e=>setCategory(e.target.value)}
+                        value={todoItem.category}
+                        onChange={(event: React.FormEvent<HTMLSelectElement>) => {
```
Comment (hunteva) [inline src/component/InputForm.tsx:27]: const eventHandler = (event: React.FormEvent<HTMLInputElement | HTMLSelectElement>) => {}

File: src/component/TodoItemDetail.tsx [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+import { TodoItemList} from '../interface/models'
+
+interface ParamTypes {
+    todoId: string
+}
+
+const TodoItemDetail = () => {
+    let history = useHistory();
+    const { todoId } = useParams<ParamTypes>()
+    const todoItem = useSelector((state: TodoItemList) => state[todoId])
```
Comment (hunteva) [inline src/component/TodoItemDetail.tsx:12]: Equality comparison

File: src/component/TodoList.tsx [hunk 14]
```diff
@@ -1,18 +1,16 @@
...
+const TodoList = () => {
     const allTodos = useSelector((state: RootState) => state)
-    const allTodoObj: ITodoItemObject[] = Object.values(allTodos)
+    const allTodoObj: TodoItem[] = Object.values(allTodos)
     const dispatch = useDispatch()
-    console.log(allTodoObj)
     return(
         <div>
-            <button onClick={() => dispatch(deleteAllTodos(allTodoObj))}>Deleted selected</button>
+            <button onClick={() => dispatch(deleteAllTodos(null))}>Deleted selected</button>
```
Comment (hunteva) [inline src/component/TodoList.tsx:13]: no passing null

File: src/redux/index.tsx [hunk 15]
```diff
@@ -1,19 +1,25 @@
...
 
 const reducer = createReducer(initialTodoList, {
-    [addTodo.type]: produce((draft, action) => {
-        draft[action.payload.id] = action.payload
-    }),
-    [deleteTodo.type]: produce((draft, action) => {delete draft[action.payload]}),
-    [deleteAllTodos.type]: produce((draft, action) => {action.payload.map((todo: { id: string }) => delete draft[todo.id])}) //don't know how to empty an object 
+    [addTodo.type]:  (state: TodoItemList, action: PayloadAction<TodoItem>) => {
+        const id = uuidv1()
+        action.payload.id = id
```
Comment (hunteva) [inline src/redux/index.tsx:14]: Changing arguments

File: src/redux/index.tsx [hunk 16]
```diff
@@ -1,19 +1,25 @@
...
 
 const initialTodoList = {}
 
 const reducer = createReducer(initialTodoList, {
-    [addTodo.type]: produce((draft, action) => {
-        draft[action.payload.id] = action.payload
-    }),
-    [deleteTodo.type]: produce((draft, action) => {delete draft[action.payload]}),
-    [deleteAllTodos.type]: produce((draft, action) => {action.payload.map((todo: { id: string }) => delete draft[todo.id])}) //don't know how to empty an object 
+    [addTodo.type]:  (state: TodoItemList, action: PayloadAction<TodoItem>) => {
```
Comment (hunteva) [inline src/redux/index.tsx:12]: (state: TodoItemList, { payload }: PayloadAction<TodoItem>)

File: src/redux/index.tsx [hunk 17]
```diff
@@ -1,19 +1,25 @@
-import { createAction, createReducer } from '@reduxjs/toolkit'
-import produce from 'immer'
-import { ITodoItem, ITodoItemObject} from '../interface/models'
+import { createAction, createReducer, PayloadAction} from '@reduxjs/toolkit'
+import { v1 as uuidv1 } from 'uuid'
+import { TodoItem, TodoItemList } from '../interface/models'
 
-export const addTodo = createAction<ITodoItemObject>('add_todo')
+export const addTodo = createAction<TodoItem>('add_todo')
```
Comment (hunteva) [inline src/redux/index.tsx:5]: export type AddTodoAction = ReturnType<typeof addTodo>;

**Pull request closed without merging**